pydantic/main.py

```diff
@@ -652,10 +652,27 @@ def _calculate_keys(
         return keys
 
     def __eq__(self, other: Any) -> bool:
-        if isinstance(other, BaseModel):
-            return self.model_dump() == other.model_dump()
-        else:
-            return self.model_dump() == other
+        if not isinstance(other, BaseModel):
+            return False
+
+        # When comparing instances of generic types for equality, as long as all field values are equal,
+        # only require their generic origin types to be equal, rather than exact type equality.
+        # This prevents headaches like MyGeneric(x=1) != MyGeneric[Any](x=1).
+        self_type = getattr(self, '__pydantic_generic_origin__', None) or self.__class__
+        other_type = getattr(other, '__pydantic_generic_origin__', None) or other.__class__
+
+        if self_type != other_type:
+            return False
+
+        if self.__dict__ != other.__dict__:
+            return False
```

Comment:
Previously I had this as a check of a tuple, but that was repeatably about 10% slower in the case where the types were different. The difference when the types were the same was negligible.